File: ros2_ws/rti_workspace/5.2.3/examples/connext_dds/cs/hello_world_request_reply/src/PrimeNumberRequester.cs
/* ****************************************************************************
 *         (c) Copyright, Real-Time Innovations, All rights reserved.       
 *                                                                          
 *         Permission to modify and use for internal purposes granted.      
 * This software is provided "as is", without warranty, express or implied. 
 *                                                                          
 * ****************************************************************************
 */
using System;
using System.Collections.Generic;

using DDS;
using RTI.Connext.RequestReply;
using RTI.Connext.Infrastructure;

namespace RequestReplyExample
{
    class PrimeNumberRequester : IDisposable
    {
        // --- Members: -------------------------------------------------------

        private DomainParticipant participant;
        private Requester<PrimeNumberRequest, PrimeNumberReply> requester;

        // --- Constructor: ---------------------------------------------------

        public PrimeNumberRequester(int domainId)
        {

            // Create participant
            participant = DomainParticipantFactory.get_instance()
                .create_participant(
                    domainId,
                    DomainParticipantFactory.PARTICIPANT_QOS_DEFAULT,
                    null, StatusMask.STATUS_MASK_NONE);

            if (participant == null)
            {
                throw new SystemException("Error creating participant");
            }

            // Create replier with that participant and:
            // - The DDS TypeSupport for the request and reply types
            // - A service name
            // - (Optionally) a QoS profile, in USER_QOS_PROFILES.xml
            requester = new Requester<PrimeNumberRequest, PrimeNumberReply>(
                    new RequesterParams(
                        participant,
                        PrimeNumberRequestTypeSupport.get_instance(),
                        PrimeNumberReplyTypeSupport.get_instance())
                    .SetServiceName("PrimeCalculator")
                    .SetQosProfile("RequestReplyExampleProfiles",
                            "RequesterExampleProfile")
            );
        }

        // --- Dispose: -------------------------------------------------------

        public void Dispose()
        {
            if (requester != null)
            {
                requester.Dispose();
                requester = null;
            }

            if (participant != null)
            {
                participant.delete_contained_entities();
                DomainParticipantFactory.get_instance().
                    delete_participant(ref participant);
                participant = null;
            }
        }

        // --- Run requester: -------------------------------------------------

        public void PrintPrimeNumbers(int n, int primesPerReply)
        {
                
            Duration_t MAX_WAIT = new Duration_t();
            MAX_WAIT.sec = 20;
            MAX_WAIT.nanosec = 0;

            // Create a request sample
            WriteSample<PrimeNumberRequest> request =
                requester.CreateRequestSample();

            request.Data.n = n;
            request.Data.primes_per_reply = primesPerReply;

            //
            // Send the request
            //
            requester.SendRequest(request);

            bool inProgress = true;
            while (inProgress)
            {
                //
                // Receive replies (one or more at a time)
                //
                // We read a collection of loaned samples that
                // is returned (Disposed) to the middleware
                // when the application exits the using block.
                //
                using (LoanedSamples<PrimeNumberReply> replies =
                        requester.ReceiveReplies(MAX_WAIT))
                {

                    if (replies.Count == 0)
                    {
                        // The Requester doesn't report a timeouts as exceptions
                        // but as a empty collection instead.
                        throw new TimeoutException(
                            "Timed out waiting for replies");
                    }

                    // Iterate through the sample collection
                    foreach (Sample<PrimeNumberReply> reply in replies)
                    {
                        if (!reply.Info.valid_data)
                        {
                            continue;
                        }

                        // Print the prime numbers we receive
                        for (int i = 0; i < reply.Data.primes.length; i++)
                        {
                            int prime = reply.Data.primes.get_at(i);
                            Console.Write(prime + " ");
                        }

                        if (reply.Data.status !=
                                PrimeNumberCalculationStatus.REPLY_IN_PROGRESS)
                        {
                            inProgress = false;
                            if (reply.Data.status ==
                                    PrimeNumberCalculationStatus.REPLY_ERROR)
                            {
                                throw new SystemException("Error in Replier");
                            }
                            else
                            { // status == COMPLETED
                                Console.WriteLine("DONE");
                            }
                        }
                        Console.WriteLine();
                    }
                }
            }
        }

        // --- Application execution: -----------------------------------------

        private static void PrintFormat()
        {
            Console.WriteLine("PrimeNumberRequester");
            Console.WriteLine(
                "Sends a request to calculate the prime numbers <= n");
            Console.WriteLine(
                "Parameters: <n> [<primes_per_reply>=5] [<domain_id>=0]");
        }

        public static void Main(string[] args)
        {
            int n;
            int primesPerReply = 5;
            int domainId = 0;

            // Parse arguments
            if (args.Length == 0)
            {
                PrintFormat();
                return;
            }

            try
            {
                n = Int32.Parse(args[0]);
            }
            catch (FormatException)
            {
                Console.WriteLine(
                    "First argument, n, must be a positive integer");
                PrintFormat();
                return;
            }

            if (args.Length > 1)
            {
                try
                {
                    primesPerReply = Int32.Parse(args[1]);
                }
                catch (FormatException)
                {
                    Console.WriteLine("Invalid domain id");
                    PrintFormat();
                    return;
                }
            }

            if (args.Length > 2)
            {
                try
                {
                    domainId = Int32.Parse(args[2]);
                }
                catch (FormatException)
                {
                    Console.WriteLine("Invalid domain id");
                    PrintFormat();
                    return;
                }
            }

            /* Uncomment this to turn on additional logging
             NDDS.ConfigLogger.get_instance().set_verbosity(
                NDDS.LogVerbosity.NDDS_CONFIG_LOG_VERBOSITY_WARNING);             
             */

            Console.WriteLine(
                "PrimeNumberRequester: Sending a request to calculate the " +
                "prime numbers <= " + n + " in sequences of " + primesPerReply +
                " or less elements (on domain " + domainId + ")");

            // Run the example
            using (PrimeNumberRequester primeNumberReplier =
                    new PrimeNumberRequester(domainId))
            {
                try
                {
                    primeNumberReplier.PrintPrimeNumbers(n, primesPerReply);
                }
                catch (System.Exception ex)
                {
                    Console.WriteLine("Exception: " + ex.Message);
                }
            }
        }
    }
}

File: rti_connext_dds-5.2.3/resource/template/rti_workspace/examples/connext_dds/java/hello_world_request_reply/src/com/rti/example/primes/PrimeNumberRequester.java
// ****************************************************************************
//         (c) Copyright, Real-Time Innovations, All rights reserved.       
//                                                                          
//         Permission to modify and use for internal purposes granted.      
// This software is provided "as is", without warranty, express or implied. 
//                                                                          
// ****************************************************************************

package com.rti.example.primes;

import java.util.concurrent.TimeoutException;

import com.rti.connext.infrastructure.Sample;
import com.rti.connext.infrastructure.WriteSample;
import com.rti.connext.requestreply.Requester;
import com.rti.connext.requestreply.RequesterParams;
import com.rti.dds.domain.DomainParticipant;
import com.rti.dds.domain.DomainParticipantFactory;
import com.rti.dds.infrastructure.Duration_t;
import com.rti.dds.infrastructure.StatusKind;
import com.rti.example.primes.idl.PrimeNumberCalculationStatus;
import com.rti.example.primes.idl.PrimeNumberReply;
import com.rti.example.primes.idl.PrimeNumberReplyTypeSupport;
import com.rti.example.primes.idl.PrimeNumberRequest;
import com.rti.example.primes.idl.PrimeNumberRequestTypeSupport;

public class PrimeNumberRequester {
    
    private DomainParticipant participant;
    private Requester<PrimeNumberRequest, PrimeNumberReply> requester;
    
    private final static Duration_t MAX_WAIT = new Duration_t(20, 0);
    
    public PrimeNumberRequester(int domainId) {
        
        // Create participant
        participant = DomainParticipantFactory.get_instance()
                .create_participant(
                        domainId, 
                        DomainParticipantFactory.PARTICIPANT_QOS_DEFAULT, 
                        null, StatusKind.STATUS_MASK_NONE);
        
        if (participant == null) {
            throw new RuntimeException("Participant creation failed");
        }

        // Create replier with that participant
        requester = new Requester<PrimeNumberRequest, PrimeNumberReply>(
                new RequesterParams(
                    participant, 
                    // The DDS TypeSupport for the request and reply types
                    PrimeNumberRequestTypeSupport.get_instance(), 
                    PrimeNumberReplyTypeSupport.get_instance())
                    // Same service name in Requester and Replier
                    .setServiceName("PrimeCalculator")
                    // QoS profiles defined in USER_QOS_PROFILES.xml
                    .setQosProfile("RequestReplyExampleProfiles", 
                            "RequesterExampleProfile"));
    }

    /**
     * 
     * @param n
     * @return The list of primes <= n
     * @throws TimeoutException If no prime numbers are received for the
     *         duration indicated by MAX_WAIT
     * @throws NumberFormatException If any of the received prime numbers
     *         is not a valid integer (except for the empty string that
     *         indicates the end of the sequence of primes)
     */
    public void printPrimeNumbers(int n, int primesPerReply) 
            throws TimeoutException {
        
        // Create a request sample
        WriteSample<PrimeNumberRequest> request = 
                requester.createRequestSample();
        
        request.getData().n = n;
        request.getData().primes_per_reply = primesPerReply;
        
        //
        // Send the request
        //
        requester.sendRequest(request);
        
        boolean inProgress = true;
        while(inProgress) {
          
            //
            // Receive replies (one or more at a time)
            // The iterator can access a loaned list of samples
            //
            Sample.Iterator<PrimeNumberReply> replies = 
                    requester.receiveReplies(MAX_WAIT);
            
            // When receiveReplies times out, 
            // it returns an empty reply collection
            if (!replies.hasNext()) {
                throw new TimeoutException("Timed out waiting for replies");
            }
            
            try {
                do { // Iterate through the sample list
                    Sample<PrimeNumberReply> reply = replies.next();
                    if (reply.getInfo().valid_data) {
                        
                        // Print the prime numbers we receive
                        for(Object primeObj : reply.getData().primes) {
                            int prime = (Integer) primeObj;
                            System.out.print(prime + " ");
                        }
                        
                        if (reply.getData().status != 
                                PrimeNumberCalculationStatus.REPLY_IN_PROGRESS){
                            inProgress = false;
                            if (reply.getData().status ==
                                    PrimeNumberCalculationStatus.REPLY_ERROR) {
                                throw new RuntimeException("Error in replier");
                            } else { // reply->status == COMPLETED
                                System.out.println("DONE");
                            }
                        }
                        
                        System.out.println();
                    }
                } while (replies.hasNext());

            } finally {
                // Return loaned samples
                // Tip: In Java 7, consider using a try-with-resources block
                replies.close(); 
            }
        }
    }
    
    // ------------------------------------------------------------------------
    
    private static void printFormat() {
        System.out.println("PrimeNumberRequester");
        System.out.println("Sends a request to calculate the prime numbers <= n");
        System.out.println("Parameters: <n> [<primes_per_reply>=5] [<domain_id>=0]");
    }

    // ------------------------------------------------------------------------
    
    public void close() {
        if (requester != null) {
            requester.close();
            requester = null;
        }
        
        if (participant != null) {
            participant.delete_contained_entities();
                DomainParticipantFactory.get_instance()
                    .delete_participant(participant);
            participant = null;
        }

    }
    
    // ------------------------------------------------------------------------
    
    public static void main(String[] args) throws TimeoutException {

        int n;
        int primesPerReply = 5;
        int domainId = 0;
        
        // Parse arguments
        if (args.length == 0) {
            printFormat();
            return;
        }
        
        try {
            n = Integer.parseInt(args[0]);
        } catch (NumberFormatException ex) {
            System.out.println("First argument, n, must be a positive integer");
            printFormat();
            return;
        }
        
        if (args.length > 1) {
            try {
                primesPerReply = Integer.parseInt(args[1]);
            } catch (NumberFormatException ex) {
                System.out.println("Invalid domain id");
                printFormat();
                return;
            }
        }
        
        if (args.length > 2) {
            try {
                domainId = Integer.parseInt(args[2]);
            } catch (NumberFormatException ex) {
                System.out.println("Invalid domain id");
                printFormat();
                return;
            }
        }
        
        /* Uncomment this to turn on additional logging
        com.rti.ndds.config.Logger.get_instance().set_verbosity(
            com.rti.ndds.config.LogVerbosity.NDDS_CONFIG_LOG_VERBOSITY_WARNING);
         */
        
        System.out.println(
            "PrimeNumberRequester: Sending a request to calculate the " +
            "prime numbers <= " + n + " in sequences of " + primesPerReply +
            " or less elements (on domain " + domainId + ")");     
        
        // Run the example
        PrimeNumberRequester primeNumberReplier = 
                new PrimeNumberRequester(domainId);
        try {
            primeNumberReplier.printPrimeNumbers(n, primesPerReply);
        } finally {
            primeNumberReplier.close();
        }
    }    
}
